### ISSUE TITLE
lazy load template error Fight with Support

### ISSUE THREAD

**Issue by Constrat** [MaaAssistantArknights/MaaAssistantArknights#6584]
### Before raising the issue...

- [X] I understand that Issues are for feedback and problem solving, not trolling the comments section, and will provide as much information as possible to help solve the problem.
- [X] I filled in a short, clear title so that developers could quickly identify the general problem when going through the issue list. Instead of "Some suggestions", "Stuck", etc.
- [X] I have checked the latest beta update and there is no mention of the bug being fixed.

### Description

`SubTaskError` happens after `BattleQuickFormation.png` template identification.

~~This happens about 1/4 of the times.~~
If this issue arises, MAA will overwrite the `"Visit.LastCreditFightTaskTime"` to a later date / time,
rendering it impossible to retry in case of an error, without manually deleting the gui.json.baks and removing the code line `"Visit.LastCreditFightTaskTime": "xxxx/xx/xx xx:xx:xx"`

### Log files

```
[2023-09-29 21:53:53.831][INF][Px244][Tx24c] {"class":"asst::ProcessTask","details":{"cur_retry":0,"retry_times":0,"to_be_recognized":["BattleStartPre"]},"first":["BattleStartPre"],"pre_task":"","subtask":"ProcessTask","taskchain":"Copilot","taskid":1}
[2023-09-29 21:53:54.249][INF][Px244][Tx24c] Call ` D:\Downloads\MAA_Y\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | nc -w 3 10.0.2.3 61984" ` ret 0 , cost 417 ms , stdout size: 0 , socket size: 8294412
[2023-09-29 21:53:54.262][TRC][Px244][Tx24c] asst::PipelineAnalyzer::analyze BattleStartPre
[2023-09-29 21:53:54.543][TRC][Px244][Tx24c] asst::WordOcr [{ AutoDeploy: [ 128, 0, 102, 15 ], score: 0.970918 }, { ce: [ 0, 65, 25, 19 ], score: 0.995664 }, { Start: [ 178, 58, 54, 28 ], score: 0.999492 }, { : [ 5, 95, 8, 11 ], score: 0.000000 }, { 2: [ 154, 92, 20, 11 ], score: 0.407774 }, { -: [ 214, 95, 8, 14 ], score: 0.595777 }] by OCR Pipeline , cost 280 ms
[2023-09-29 21:53:54.543][TRC][Px244][Tx24c] Proced [{ Start: [ 1178, 658, 54, 28 ], score: 0.999492 }]
[2023-09-29 21:53:54.543][INF][Px244][Tx24c] Assistant::append_callback | SubTaskStart {"class":"asst::ProcessTask","details":{"action":"ClickSelf","algorithm":"OcrDetect","exec_times":1,"max_times":2147483647,"task":"BattleStartPre"},"first":["BattleStartPre"],"pre_task":"","subtask":"ProcessTask","taskchain":"Copilot","taskid":1,"uuid":"5c67b16a8af5066f"}
[2023-09-29 21:53:54.543][TRC][Px244][Tx24c] minitouch click: (1806, 1003)
[2023-09-29 21:53:54.644][TRC][Px244][Tx24c] ready to sleep 1000
[2023-09-29 21:53:55.644][TRC][Px244][Tx24c] end of sleep 1000
[2023-09-29 21:53:55.644][INF][Px244][Tx24c] Assistant::append_callback | SubTaskCompleted {"class":"asst::ProcessTask","details":{"action":"ClickSelf","algorithm":"OcrDetect","exec_times":1,"max_times":2147483647,"task":"BattleStartPre"},"first":["BattleStartPre"],"pre_task":"","subtask":"ProcessTask","taskchain":"Copilot","taskid":1,"uuid":"5c67b16a8af5066f"}
[2023-09-29 21:53:55.644][TRC][Px244][Tx24c] asst::ProcessTask::_run | leave, 1813 ms
[2023-09-29 21:53:55.644][TRC][Px244][Tx24c] ready to sleep 500
[2023-09-29 21:53:56.145][TRC][Px244][Tx24c] end of sleep 500
[2023-09-29 21:53:56.145][INF][Px244][Tx24c] Assistant::append_callback | SubTaskStart {"class":"asst::BattleFormationTask","details":{},"subtask":"BattleFormationTask","taskchain":"Copilot","taskid":1,"uuid":"5c67b16a8af5066f"}
[2023-09-29 21:53:56.145][TRC][Px244][Tx24c] asst::BattleFormationTask::_run | enter
[2023-09-29 21:53:56.145][INF][Px244][Tx24c] Assistant::append_callback | SubTaskExtraInfo {"class":"asst::BattleFormationTask","details":{"formation":["先锋"]},"subtask":"BattleFormationTask","taskchain":"Copilot","taskid":1,"uuid":"5c67b16a8af5066f","what":"BattleFormation"}
[2023-09-29 21:53:56.145][TRC][Px244][Tx24c] asst::ProcessTask::_run | enter
[2023-09-29 21:53:56.145][INF][Px244][Tx24c] {"class":"asst::ProcessTask","details":{"cur_retry":0,"retry_times":0,"to_be_recognized":["BattleQuickFormation"]},"first":["BattleQuickFormation"],"pre_task":"","subtask":"ProcessTask","taskchain":"Copilot","taskid":1}
[2023-09-29 21:53:56.555][INF][Px244][Tx24c] Call ` D:\Downloads\MAA_Y\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | nc -w 3 10.0.2.3 61984" ` ret 0 , cost 409 ms , stdout size: 0 , socket size: 8294412
[2023-09-29 21:53:56.568][TRC][Px244][Tx24c] asst::PipelineAnalyzer::analyze BattleQuickFormation
[2023-09-29 21:53:56.568][INF][Px244][Tx24c] asst::TemplResource::get_templ lazy load BattleQuickFormation.png
[2023-09-29 21:53:56.579][TRC][Px244][Tx24c] asst::ProcessTask::_run | leave, 433 ms
[2023-09-29 21:53:56.579][TRC][Px244][Tx24c] ready to sleep 500
[2023-09-29 21:53:57.079][TRC][Px244][Tx24c] end of sleep 500
[2023-09-29 21:53:57.079][INF][Px244][Tx24c] Assistant::append_callback | SubTaskError {"class":"asst::ProcessTask","details":{},"first":["BattleQuickFormation"],"pre_task":"","subtask":"ProcessTask","taskchain":"Copilot","taskid":1,"uuid":"5c67b16a8af5066f"}
[2023-09-29 21:53:57.079][TRC][Px244][Tx24c] asst::ProcessTask::on_run_fails | enter
[2023-09-29 21:53:57.079][TRC][Px244][Tx24c] asst::ProcessTask::on_run_fails | leave, 0 ms
[2023-09-29 21:53:57.079][TRC][Px244][Tx24c] asst::BattleFormationTask::_run | leave, 933 ms
[2023-09-29 21:53:57.079][TRC][Px244][Tx24c] ready to sleep 500
[2023-09-29 21:53:57.580][TRC][Px244][Tx24c] end of sleep 500
[2023-09-29 21:53:57.580][INF][Px244][Tx24c] Assistant::append_callback | SubTaskError {"class":"asst::BattleFormationTask","details":{},"subtask":"BattleFormationTask","taskchain":"Copilot","taskid":1,"uuid":"5c67b16a8af5066f"}
[2023-09-29 21:53:57.580][TRC][Px244][Tx24c] ready to sleep 500
[2023-09-29 21:53:58.081][TRC][Px244][Tx24c] end of sleep 500
[2023-09-29 21:53:58.081][TRC][Px244][Tx24c] asst::ProcessTask::_run | enter
[2023-09-29 21:53:58.081][INF][Px244][Tx24c] {"class":"asst::ProcessTask","details":{"cur_retry":0,"retry_times":20,"to_be_recognized":["MallBegin"]},"first":["MallBegin"],"pre_task":"","subtask":"ProcessTask","taskchain":"Mall","taskid":1}
[2023-09-29 21:53:58.081][INF][Px244][Tx24c] Assistant::append_callback | SubTaskStart {"class":"asst::ProcessTask","details":{"action":"DoNothing","algorithm":"JustReturn","exec_times":1,"max_times":2147483647,"task":"MallBegin"},"first":["MallBegin"],"pre_task":"","subtask":"ProcessTask","taskchain":"Mall","taskid":1,"uuid":"5c67b16a8af5066f"}
[2023-09-29 21:53:58.081][INF][Px244][Tx24c] Assistant::append_callback | SubTaskCompleted {"class":"asst::ProcessTask","details":{"action":"DoNothing","algorithm":"JustReturn","exec_times":1,"max_times":2147483647,"task":"MallBegin"},"first":["MallBegin"],"pre_task":"","subtask":"ProcessTask","taskchain":"Mall","taskid":1,"uuid":"5c67b16a8af5066f"}
[2023-09-29 21:53:58.081][TRC][Px244][Tx24c] ready to sleep 500
[2023-09-29 21:53:58.581][TRC][Px244][Tx24c] end of sleep 500
[2023-09-29 21:53:58.581][INF][Px244][Tx24c] {"class":"asst::ProcessTask","details":{"cur_retry":0,"retry_times":20,"to_be_recognized":["Mall@ReturnTo","Mall","Mall@CloseAnno","Mall@CloseAnnoTexas","Mall@TodaysSupplies","Mall@NotDownloadVoice","Mall@StartUp","Mall@FromStageSN"]},"first":["MallBegin"],"pre_task":"MallBegin","subtask":"ProcessTask","taskchain":"Mall","taskid":1}
[2023-09-29 21:53:58.973][INF][Px244][Tx24c] Call ` D:\Downloads\MAA_Y\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | nc -w 3 10.0.2.3 61984" ` ret 0 , cost 391 ms , stdout size: 0 , socket size: 8294412
[2023-09-29 21:53:58.984][TRC][Px244][Tx24c] asst::PipelineAnalyzer::analyze Mall@ReturnTo
[2023-09-29 21:53:58.986][TRC][Px244][Tx24c] match_templ | Return.png score: 0.983232 rect: [ 18, 16, 142, 45 ] roi: [ 0, 0, 180, 80 ]
[2023-09-29 21:53:58.986][INF][Px244][Tx24c] Assistant::append_callback | SubTaskStart {"class":"asst::ProcessTask","details":{"action":"ClickSelf","algorithm":"MatchTemplate","exec_times":1,"max_times":2147483647,"task":"ReturnTo"},"first":["MallBegin"],"pre_task":"MallBegin","subtask":"ProcessTask","taskchain":"Mall","taskid":1,"uuid":"5c67b16a8af5066f"}
[2023-09-29 21:53:58.987][TRC][Px244][Tx24c] minitouch click: (124, 51)
[2023-09-29 21:53:59.088][INF][Px244][Tx24c] Assistant::append_callback | SubTaskCompleted {"class":"asst::ProcessTask","details":{"action":"ClickSelf","algorithm":"MatchTemplate","exec_times":1,"max_times":2147483647,"task":"ReturnTo"},"first":["MallBegin"],"pre_task":"MallBegin","subtask":"ProcessTask","taskchain":"Mall","taskid":1,"uuid":"5c67b16a8af5066f"}
[2023-09-29 21:53:59.088][TRC][Px244][Tx24c] ready to sleep 500
[2023-09-29 21:53:59.589][TRC][Px244][Tx24c] end of sleep 500

```

**Comment by Constrat** [comment]
```
[2023-09-30 13:12:50.568][INF][Px260c][Tx2ca0] {"class":"asst::ProcessTask","details":{"cur_retry":0,"retry_times":0,"to_be_recognized":["BattleQuickFormation"]},"first":["BattleQuickFormation"],"pre_task":"","subtask":"ProcessTask","taskchain":"Copilot","taskid":17}
[2023-09-30 13:12:51.415][INF][Px260c][Tx2ca0] Call ` D:\Downloads\MAA_Y\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | gzip -1" ` ret 0 , cost 847 ms , stdout size: 4037114 , socket size: 0
[2023-09-30 13:12:51.480][TRC][Px260c][Tx2ca0] asst::PipelineAnalyzer::analyze BattleQuickFormation
[2023-09-30 13:12:51.480][INF][Px260c][Tx2ca0] asst::TemplResource::get_templ lazy load BattleQuickFormation.png
[2023-09-30 13:12:51.504][TRC][Px260c][Tx2ca0] match_templ | BattleQuickFormation.png score: 0.732902 rect: [ 866, 18, 78, 46 ] roi: [ 816, 0, 404, 149 ]
[2023-09-30 13:12:51.504][TRC][Px260c][Tx2ca0] asst::ProcessTask::_run | leave, 936 ms
[2023-09-30 13:12:51.504][TRC][Px260c][Tx2ca0] ready to sleep 500
[2023-09-30 13:12:52.005][TRC][Px260c][Tx2ca0] end of sleep 500
[2023-09-30 13:12:52.005][INF][Px260c][Tx2ca0] Assistant::append_callback | SubTaskError {"class":"asst::ProcessTask","details":{},"first":["BattleQuickFormation"],"pre_task":"","subtask":"ProcessTask","taskchain":"Copilot","taskid":17,"uuid":"5c67b16a8af5066f"}
[2023-09-30 13:12:52.005][TRC][Px260c][Tx2ca0] asst::ProcessTask::on_run_fails | enter
[2023-09-30 13:12:52.005][TRC][Px260c][Tx2ca0] asst::ProcessTask::on_run_fails | leave, 0 ms
[2023-09-30 13:12:52.005][TRC][Px260c][Tx2ca0] asst::BattleFormationTask::_run | leave, 1437 ms
[2023-09-30 13:12:52.005][TRC][Px260c][Tx2ca0] ready to sleep 500
[2023-09-30 13:12:52.506][TRC][Px260c][Tx2ca0] end of sleep 500
[2023-09-30 13:12:52.506][INF][Px260c][Tx2ca0] Assistant::append_callback | SubTaskError {"class":"asst::BattleFormationTask","details":{},"subtask":"BattleFormationTask","taskchain":"Copilot","taskid":17,"uuid":"5c67b16a8af5066f"}
[2023-09-30 13:12:52.506][TRC][Px260c][Tx2ca0] ready to sleep 500
[2023-09-30 13:12:53.007][TRC][Px260c][Tx2ca0] end of sleep 500
[2023-09-30 13:12:53.007][TRC][Px260c][Tx2ca0] asst::ProcessTask::_run | enter
[2023-09-30 13:12:53.007][INF][Px260c][Tx2ca0] {"class":"asst::ProcessTask","details":{"cur_retry":0,"retry_times":20,"to_be_recognized":["MallBegin"]},"first":["MallBegin"],"pre_task":"","subtask":"ProcessTask","taskchain":"Mall","taskid":17}
[2023-09-30 13:12:53.007][INF][Px260c][Tx2ca0] Assistant::append_callback | SubTaskStart {"class":"asst::ProcessTask","details":{"action":"DoNothing","algorithm":"JustReturn","exec_times":1,"max_times":2147483647,"task":"MallBegin"},"first":["MallBegin"],"pre_task":"","subtask":"ProcessTask","taskchain":"Mall","taskid":17,"uuid":"5c67b16a8af5066f"}
[2023-09-30 13:12:53.007][INF][Px260c][Tx2ca0] Assistant::append_callback | SubTaskCompleted {"class":"asst::ProcessTask","details":{"action":"DoNothing","algorithm":"JustReturn","exec_times":1,"max_times":2147483647,"task":"MallBegin"},"first":["MallBegin"],"pre_task":"","subtask":"ProcessTask","taskchain":"Mall","taskid":17,"uuid":"5c67b16a8af5066f"}
[2023-09-30 13:12:53.007][TRC][Px260c][Tx2ca0] ready to sleep 500
[2023-09-30 13:12:53.508][TRC][Px260c][Tx2ca0] end of sleep 500
```
Happened again today. This time the score was displayed and it's less than the default 0.8 required. I honestly have no idea why it was that low, considering the RECT is in the perfect place.

**Comment by Constrat** [comment]
After more and more testing I've noticed a pattern and I can 10/10 repeat this issue:
If the template is not loaded before the recognition, it will not work. If the template was loaded before (for example from the time of the first error) it will work and detect the .png correctly (usually with a score of `≥0.98`)

**Comment by Constrat** [comment]
For some reason, this issue doesn't affect the Official / BiliBili client. @zzyyyl In the 4.24.0 main release, no changes were made to the tasks.json algorithms, correct?

**Comment by zzyyyl** [comment]
> In the 4.24.0 main release, no changes were made to the tasks.json algorithms, correct?

yep, #6478 was merged on _Sep 29_, and `v4.24.0` was released on _Sep 26_

**Comment by Constrat** [comment]
So that's not it either. I can't understand why the Official doesn't have the issue, while EN does. Looking at the scores, it feels like it gets loaded incorrectly maybe?

**Comment by zzyyyl** [comment]
Maybe we can try debugging it with a breakpoint in `TemplResource.cpp` - `asst::TemplResource::get_templ`

**Comment by Constrat** [comment]
I can definitely try, even though I feel like this is new. I've been using the Combat Support Friends for quite some time now.

**Comment by Constrat** [comment]
I feel like this will take ages xD... That breakpoint pretty much blocks everything.

**Comment by zzyyyl** [comment]
you can add
```cpp
if (name == "BattleQuickFormation.png") {
    Log.Trace("...");
}
```
and breakpoint at the `Log.Trace` line. ：）

**Comment by Constrat** [comment]
Thank you, I'm not really used to C++, especially at this level of complexity :P

**Comment by Constrat** [comment]
It went in `Matcher::analyze()` but as expected it skipped all the IFs because the threshold is too low.
Looking at this>m_result>rect it doesn't look good as it's x=0, y=0, width=0

**Comment by Constrat** [comment]
The correct path is selected. Aaaaaaaaaaand this time it went inside the correct if, because I'm noticing the correcy RECT.

**Comment by Constrat** [comment]
Ok... It did it again correctly, without the added debug code.

**Comment by Constrat** [comment]
I can't reproduce it anymore 👀

**Comment by Constrat** [comment]
I have the CN and the EN side to side, with the same type of instance. CN entered the Quick formation correctly, EN didn't recognize the Button

**Comment by Constrat** [comment]
@zzyyyl After hitting my head on everything it seems like MAA doesn't like ATSC textures Hard decoding... I've tried on the same Instance with a different computer and it worked but I had soft decoding on. 
On the server computer, changed the textures to off, and it worked first try...

**Comment by Constrat** [comment]
@zzyyyl made a mistake. This has not been foxed with the ATSC

**Comment by Constrat** [comment]
Ok, I've made a new BattleQuickFormation.png (a bit smaller) and placed it in `YostarEN\resource` and now it seems to be working.

**Comment by Constrat** [comment]
Unfortunately I don't have other Overseas accounts to test this. But I'm wondering if the same issue happens on JP or KR, considering they should be working in the same exact way.
I don't consider this a fix. With 4.24.0 `BattleQuickFormation.png` detection was working properly, but I'm too lazy to dig in deeper.

**Comment by Constrat** [comment]
I'm going crazy. I've tried everything but this issue always comes back:
I've created an ad-hoc task in YostarEN:
<details><summary>Task</summary>
<p>

```json
  "BattleQuickFormation": {
        "template": "BattleQuickFormationEN.png",
        "action": "ClickSelf",
        "roi": [
            816,
            0,
            404,
            149
        ],
        "postDelay": 1000,
        "next": [
            "BattleQuickFormationClear",
            "BattleQuickFormation"
        ]
    },
```
</p>
</details> 

Copying and renaming the same exact template to BattleQuickFormationEN.png, only to have the same exact issue: 
<details><summary>Log</summary>
<p>

```js
[2023-10-09 18:14:29.279][INF][Px2d90][Tx2e78] {"class":"asst::ProcessTask","details":{"cur_retry":0,"retry_times":0,"to_be_recognized":["BattleQuickFormation"]},"first":["BattleQuickFormation"],"pre_task":"","subtask":"ProcessTask","taskchain":"Copilot","taskid":2}
[2023-10-09 18:14:30.149][INF][Px2d90][Tx2e78] Call ` D:\Downloads\MAA_Y\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | gzip -1" ` ret 0 , cost 869 ms , stdout size: 3578895 , socket size: 0
[2023-10-09 18:14:30.208][TRC][Px2d90][Tx2e78] asst::PipelineAnalyzer::analyze BattleQuickFormation
[2023-10-09 18:14:30.208][INF][Px2d90][Tx2e78] asst::TemplResource::get_templ lazy load BattleQuickFormationEN.png
[2023-10-09 18:14:30.223][TRC][Px2d90][Tx2e78] asst::ProcessTask::_run | leave, 944 ms
[2023-10-09 18:14:30.223][TRC][Px2d90][Tx2e78] ready to sleep 500
[2023-10-09 18:14:30.724][TRC][Px2d90][Tx2e78] end of sleep 500
[2023-10-09 18:14:30.724][INF][Px2d90][Tx2e78] Assistant::append_callback | SubTaskError {"class":"asst::ProcessTask","details":{},"first":["BattleQuickFormation"],"pre_task":"","subtask":"ProcessTask","taskchain":"Copilot","taskid":2,"uuid":"5c67b16a8af5066f"}
[2023-10-09 18:14:30.724][TRC][Px2d90][Tx2e78] asst::ProcessTask::on_run_fails | enter
[2023-10-09 18:14:30.724][TRC][Px2d90][Tx2e78] asst::ProcessTask::on_run_fails | leave, 0 ms
[2023-10-09 18:14:30.724][TRC][Px2d90][Tx2e78] asst::BattleFormationTask::_run | leave, 1445 ms
[2023-10-09 18:14:30.724][TRC][Px2d90][Tx2e78] ready to sleep 500
[2023-10-09 18:14:31.225][TRC][Px2d90][Tx2e78] end of sleep 500
[2023-10-09 18:14:31.225][INF][Px2d90][Tx2e78] Assistant::append_callback | SubTaskError {"class":"asst::BattleFormationTask","details":{},"subtask":"BattleFormationTask","taskchain":"Copilot","taskid":2,"uuid":"5c67b16a8af5066f"}
[2023-10-09 18:14:31.225][TRC][Px2d90][Tx2e78] ready to sleep 500
[2023-10-09 18:14:31.725][TRC][Px2d90][Tx2e78] end of sleep 500
[2023-10-09 18:14:31.725][TRC][Px2d90][Tx2e78] asst::ProcessTask::_run | enter
```
</p>
</details>

**Comment by Constrat** [comment]
Like @zzyyyl to try adding a breakpoint. In the "DEBUG VERSION" from Visual Studio, the issue doesn't happen, maybe because the debug version is a bit slower(?)

**Comment by zzyyyl** [comment]
![image](https://github.com/MaaAssistantArknights/MaaAssistantArknights/assets/74587068/37d4455c-1d2b-419d-a69e-61909ae7572d)

**Comment by Constrat** [comment]
I've tried, cmake still puts DEBUG VERSION, this way the issue doesn't happen

**Comment by zzyyyl** [comment]
add debug screenshots: a3f04669db1a241d542512d2507ca67815732c1b

**Comment by Constrat** [comment]
I'll try it asap

**Comment by zzyyyl** [comment]
Check the `debug/others` screenshots and see if there are shadows or other distractions.

**Comment by Constrat** [comment]
> Using the latest nightly, with the added debug the image is exactly what we expect it to: ![2023-10-11_16-32-47-573_raw](https://user-images.githubusercontent.com/56174894/274311630-6cc53a4b-10b6-46af-b822-47ccfb06e24f.png) But that's not promising.
It's this one... It's perfect

**Comment by Constrat** [comment]
So the debug image that we see is not what gets analyzed?

**Comment by zzyyyl** [comment]
> So the debug image that we see is not what gets analyzed?

I think so

**Comment by zzyyyl** [comment]
I'll see if I can save the analyzed image directly.

**Comment by Constrat** [comment]
I dont' think that makes a difference, as you can see from the edited post there are two happening where one time it doesn't detect, the other it does, and size's the same

**Comment by zzyyyl** [comment]
I think the template mismatch is simply due to shadowing when "Submitting feedback to the neural network", and I've increased the max retry times to 3.

**Comment by Constrat** [comment]
I haven't debugged, but it seems like With @zzyyyl latest commit it now "works".
I'll see at 13:00 when there's the reset if automatically works

**Comment by Constrat** [comment]
Fixed in 734cb19
@zzyyyl 🙏🏻
